articles/101_build_tool.md

```diff
@@ -0,0 +1,261 @@
+Extensions don't even have to be integrated into the Python package containing the core logic of the build tool but can easily be provided by additional Python packages.
+This approach will not only foster a modular design and promote clear interfaces but enable external contributions without requiring them to be integrated in a single monolithic package.
+
+Several well known software principles apply:
+
+- Separation of concerns
+- Single Responsibility principle
+- Principle of Least Knowledge
+- Don’t repeat yourself
+- Keep it stupid simple
+- "Not paying for what you don't use"
+
+### Extension Points
+
+The following items are possible extension points to provide custom functionality:
+
+- contribute `verbs` to the command line tool (e.g. `build`, `test`)
+- contribute command line options for specific features (e.g. `build`, `test`)
+- discovery of packages (e.g. recursively crawling a workspace)
+- identification of packages and their meta information (e.g. from a `package.xml` file)
+- process a package (e.g. build a CMake package, test a Python package)
+- execution model (e.g. sequential processing, parallel processing)
+- output handling (e.g. console output, logfiles, status messages, notifications)
+- setup the environment (e.g. `sh`, `bash`, `bat`)
+- completion (e.g. `bash`, `Powershell`)
+
+## Possible Approaches
+
+In terms of flexibility neither of the existing build tools can already support the superset of features described in this article.
+There are two different paths possible to reach the goal of a universal build tool:
+
+### Evolve catkin_tools
+
+One approach is to incrementally evolve one of the existing tools to satisfy the described goals.
+Since `catkin_tools` is in many aspects the most complete build tool it should be one being evolved.
+While `ament_tools` has a few features `catkin_tools` currently lacks (e.g. plain CMake support without a manifest, Windows support) the feature richness of `catkin_tools` makes it a better starting point.
+
+The following items are highlighting some of the necessary efforts (not a complete list):
+
+- Refactor the software architecture of the existing code base to support the flexibility sketched by the extension points listed above.
+- Move `catkin` specific concepts out of the core of the build tool into a catkin specific extension (e.g. manifest format, *devel space*).
+- Support for ROS 2 which includes:
+
+  - Support for Python 3 and Windows.
+  - Support for pure Python packages as well as packages without an in-source manifest file.
+  - Environment setup of `ament` packages
+
+- Rename the tool to use a name unrelated to one build system.
+- Investigate if a feature like continued support of the *devel space* is feasible since it doesn't apply to other build system and might be complicated to separate without sacrificing usability.
```

Comment:
How is this specific to evolving catkin_tools? What would the 'from scratch' solution do w.r.t. the devel space?